x/sequencer/keeper/unbond.go

```diff
@@ -26,158 +49,130 @@ func (k Keeper) UnbondAllMatureSequencers(ctx sdk.Context, currTime time.Time) {
-		),
-	)
+func (k Keeper) unbondSequencerAndJail(ctx sdk.Context, seqAddr string) error {
+	return k.unbond(ctx, seqAddr, true)
+}
 
-	return nil
+func (k Keeper) unbondSequencer(ctx sdk.Context, seqAddr string) error {
+	return k.unbond(ctx, seqAddr, false)
 }
 
-// unbondUnbondingSequencer unbonds a sequencer that currently unbonding
-func (k Keeper) unbondUnbondingSequencer(ctx sdk.Context, seqAddr string) error {
+func (k Keeper) unbond(ctx sdk.Context, seqAddr string, jail bool) error {
 	seq, found := k.GetSequencer(ctx, seqAddr)
 	if !found {
 		return types.ErrUnknownSequencer
 	}
 
-	if seq.Status != types.Unbonding {
+	if seq.Status == types.Unbonded {
 		return errorsmod.Wrapf(
 			types.ErrInvalidSequencerStatus,
-			"sequencer status is not unbonding: got %s",
-			seq.Status.String(),
+			"sequencer status is already unbonded",
 		)
 	}
-	seqTokens := seq.Tokens
-	if !seqTokens.Empty() {
-		seqAcc, err := sdk.AccAddressFromBech32(seq.Address)
-		if err != nil {
-			return err
+	// keep the old status for updating the sequencer
+	oldStatus := seq.Status
+
+	// handle bond: tokens refunded by default, unless jail is true
+	err := k.reduceSequencerBond(ctx, &seq, seq.Tokens, jail)
+	if err != nil {
+		return errorsmod.Wrap(err, "remove sequencer bond")
+	}
+
+	/* ------------------------------ store cleanup ----------------------------- */
+	// remove from queue if unbonding
+	if oldStatus == types.Unbonding {
+		k.removeUnbondingSequencer(ctx, seq)
+	} else {
+		// remove from notice period queue if needed
+		if seq.IsNoticePeriodInProgress() {
+			k.removeNoticePeriodSequencer(ctx, seq)
 		}
 
-		err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, seqAcc, seqTokens)
-		if err != nil {
-			return err
+		// if we unbond the the proposer, remove it
+		// the caller should rotate the proposer
+		if k.isProposer(ctx, seq.RollappId, seqAddr) {
+			k.removeProposer(ctx, seq.RollappId)
+		}
+
+		// if we unbond the next proposer, we're in the middle of rotation
+		// the caller should clean the rotation state
+		if k.isNextProposer(ctx, seq.RollappId, seqAddr) {
+			k.removeNextProposer(ctx, seq.RollappId)
+		}
+	}
+	// in case the sequencer is currently reducing its bond, then we need to remove it from the decreasing bond queue
+	// all the tokens are returned, so we don't need to reduce the bond anymore
+	if bondReductions := k.getSequencerDecreasingBonds(ctx, seq.Address); len(bondReductions) > 0 {
+		for _, bondReduce := range bondReductions {
+			k.removeDecreasingBondQueue(ctx, bondReduce)
 		}
-	} else {
-		k.Logger(ctx).Error("sequencer has no tokens to unbond", "sequencer", seq.Address)
 	}
 
-	// set the status to unbonded and remove from the unbonding queue
-	seq.Status = types.Unbonded
-	seq.Tokens = sdk.Coins{}
+	if jail {
+		seq.Jailed = true
+	}
+	// set the unbonding height and time, if not already set
+	if seq.UnbondRequestHeight == 0 {
+		seq.UnbondRequestHeight = ctx.BlockHeight()
+	}
```

Comment:
this is overload of bond request height, it's kind of a fake meaning
why need to set it? should explain